README.md
# A Big Stick
A Terraria mod that contains numerous stick-themed items.

# Weapons
## Handguns
- Wooden Stickgun - 50 Wood (Made at a Work Bench)
- Demonite Stickgun - 15 Demonite Bars, 50 Wood (Made at an Anvil)
- Crimtane Stickgun - 15 Crimtane Bars, 50 Wood (Made at an Anvil)
- Hellstone Stickgun - 15 Hellstone Bars, 50 Wood (Made at an Anvil)
- Adamantite Stickgun - 12 Adamantite Bars, 50 Wood (Made at a Mythril or Orichalcum Anvil)
- Titanium Stickgun - 12 Titanium Bars, 50 Wood (Made at a Mythril or Orichalcum Anvil)
- Autotrophic Stickgun - 12 Chlorophyte Bars, 50 Wood (Made at a Mythril or Orichalcum Anvil)
- Spectre Stickgun - 12 Spectre Bars, 50 Wood (Made at a Mythril or Orichalcum Anvil)
- Lunar Stickgun - 28 Vortex Fragments, 50 Wood (Made at an Ancient Manipulator)
- Clockwork Stickgun - 1 Phantasm/Vortex Beater, 5 Luminite Bars, 50 Wood (Made at an Ancient Manipulator)
## Rifles
- Wooden Stickrifle - 50 Wood (Made at a Work Bench)
- Demonite Stickrifle - 15 Demonite Bars, 50 Wood (Made at an Anvil)
- Crimtane Stickrifle - 15 Crimtane Bars, 50 Wood (Made at an Anvil)
- Hellstone Stickrifle - 15 Hellstone Bars, 50 Wood (Made at an Anvil)
- Adamantite Stickrifle - 12 Adamantite Bars, 50 Wood (Made at a Mythril or Orichalcum Anvil)
- Titanium Stickrifle - 12 Titanium Bars, 50 Wood (Made at a Mythril or Orichalcum Anvil)
- Autotrophic Stickrifle - 12 Chlorophyte Bars, 50 Wood (Made at a Mythril or Orichalcum Anvil)
- Spectre Stickrifle - 12 Spectre Bars, 50 Wood (Made at a Mythril or Orichalcum Anvil)
- Lunar Stickrifle - 28 Vortex Fragments, 50 Wood (Made at an Ancient Manipulator)
- Clockwork Stickrifle - 1 Phantasm/Vortex Beater, 5 Luminite Bars, 50 Wood (Made at an Ancient Manipulator)
# Sticks
- 30 Sticks - 1 Wood/Palm Wood/Rich Mahogany
- 30 Boreal Sticks - 1 Boreal Wood (Inflicts Chilled)
- 30 Pearlsticks - 1 Pearlwood (Glows, ignores gravity, inflicts Slow)
- 30 Ebonsticks - 1 Ebonwood (Inflicts Cursed Inferno)
- 30 Shadesticks - 1 Shadewood (Inflicts Ichor)
- 30 Spooky Sticks - 1 Spooky Wood (33% chance to inflict Confusion)
- 100 High Velocity Sticks - 100 Sticks, 1 Cog (Goes much quicker than all other sticks, ignores gravity, made at a Workbench)
- 150 Meteorite Sticks - 1 Meteorite Bar (Pierces through 5 entities at max, but doesn't ignore invincibility frames, made at an Anvil)
- 333 Luminite Sticks - 1 Luminite Bar (Increases in damage the more targets it pierces through, doesn't ignore invincibility frames, made at an Ancient Manipulator)
- 333 Ether Sticks - 1 of any Fragment (Increases in damage the longer it travels, ignores gravity, made at an Ancient Manipulator)
- 500 Sticks of Tears - 25 Last Tears (Made at an Anvil)
- 200 Lunar Sticks - 10 Lunar Tears (Made at a Mythril Anvil)
- 200 Cosmilite Stick - 5 Cosmilite Bars (Only available if you have the Calamity Mod, made at an Ancient Manipulator)
- Greedy Stick - Dropped by Pirates (Pierces through 5 entities at max, but doesn't ignore invincibility frames, has a chance to steal money upon hitting anything but a Target Dummy)
- Sentient Stick - Dropped by Sentient Stickgun (Homes in on targets)
# NPCs
- A Sentient Stickgun - Boss summoned by Twitching Stickgun. Drops Sentient Sticks
# Other Items
- Twitching Stickgun - 1 Clockwork Stickgun, 99 Nanites (Summons a Sentient Stickgun, made at a Demon Altar)
- A Big Stick - Dropped from a Sentient Stickgun; summons Sticks from the sky
- Net of Tears - Bug Net, 10 Hellstone Bars (Mobs drop Last Tears upon death if equipped, made at an Anvil)
- Lunar Tears - Dropped by the Moon Lord
# Credits
- Atenfyr - Programming and Sprites
- Geocide - Sprites

CHANGELOG.md
##2.5.0
- Added Sticks and Meteorite Sticks to the Arms Dealer's shop.
- Changed the High Velocity Stick recipe to allow any kind of stick.
##2.4.0
- Updated the hitbox of Ether Sticks to be the same as that of all other sticks.
- Updated the Sentient Stickgun so that it can be spawned during daytime.
- Updated Sentient Sticks to adjust their velocity to home towards a target rather than completely ignoring the old velocity.
- Increased the number of Sentient Sticks that drop from the Sentient Stickgun.
- Increased the health of the Sentient Stickgun.
- Increased the size of the Sentient Stickgun.
- Changed the texture of the Sentient Stickgun to that of the Sentient Stickrifle.
- Fixed a bug where the turret spawned by the Big Stick would not despawn in multiplayer.
- Fixed a bug where Sentient Sticks would home in on invincible entities.
##2.3.1
- Fixed the recipe of Ether Sticks.
##2.3.0
- Added Ether Sticks.
##2.2.4
- Increased the drop rate of Greedy Sticks.
##2.2.3
- Increased the money dropped by hitting monsters with Greedy Sticks.
- Increased the speed of Pearlsticks.
- Made Pearlsticks glow.
- Removed the bullet spread on Spectre Stickguns and Lunar Stickguns.
- Removed gravity on High Velocity Sticks and Pearlsticks.
##2.2.2
- Made High Velocity Sticks less expensive.
- Made High Velocity Sticks somewhat visible.
##2.2.1
- Fixed a bug where Meteorite Sticks were made at a workbench rather than an anvil.
##2.2.0
- Added Meteorite Sticks and High Velocity Sticks.
- Removed the bullet spread on the Clockwork Stickgun.
##2.1.3
- Decreased the cost of Sticks of Tears.
- Fixed a bug where the rarity of Sticks of Tears was too high.
##2.1.2
- Buffed the Hellstone Stickgun and Hellstone Stickrifle.
##2.1.1
- Doubled the damage of Sticks of Tears.
##2.1.0
- Added Crimtane Stickguns and Stickrifles.
- Increased the knockback of Stickrifles.
- Increased the damage of the Big Stick.
- Reduced the duration of Cursed Inferno and Ichor inflicted from Ebonsticks and Shadesticks respectively.
- Changed Sticks of Tears to be craftable before the Wall of Flesh rather than after the Moon Lord.
##2.0.1
- Adjusted the description to be shorter.
##2.0.0
- Added nine brand new Stickguns.
- Added nine brand new Stickrifles.
- Redesigned the sprites of all Stickguns, the Big Stick, and the Sentient Stickgun, courtesy of Geocide.
- Revamped the AI of the Sentient Stickgun, including support for Calamity's Revengeance mode.
- Changed the Sentient Stickgun to drop the Big Stick.
- Changed most Sticks to ignore invincibility frames.
- Changed Greedy Sticks to always drop, regardless of whether or not the Vortex Tower has been downed.
- Changed Greedy Sticks to drop up to 20 silver rather than 2 silver.
- Reduced the cost of all Sticks.
- Gave the Clockwork Stickgun a new crafting recipe.
- Nerfed the Clockwork Stickgun.
- Removed the crafting recipe for the Big Stick and gave it a special ability.
- Removed the ability for the Clockwork Stickgun to drop from any boss bag.
- Removed the original Stickgun (without a prefix).
##1.6.0
- Added the Sentient Stickgun.
- Added the Twitching Stickgun.
- Increased the knockback of the Clockwork Stickgun.
- Changed the Clockwork Stickgun to shoot sticks in a spread.
- Removed the ability for the Clockwork Stickgun to drop from the boss bags of all hardmode bosses.
##1.5.0
- Added Cosmilite Sticks, which can only be crafted if Calamity is installed.
- Changed Lunar Sticks so that they ignore invincibility frames.
- Increased the output of the recipe for Lunar Sticks.
- Increased the droprate of the Clockwork Stickgun from boss bags.
##1.4.2
- Fixed a bug where Luminite Sticks would not pierce through enemies.
##1.4.1
- Fixed a bug where Lunar Sticks and Sticks of Tears would not pierce through enemies.
##1.4.0
- Added Lunar Tears and Lunar Sticks.
- Renamed the Bleached Stickgun to the Clockwork Stickgun.
- Changed the fire sound of the Clockwork Stickgun.
- Adjusted the Clockwork Stickgun to use more ammo.
- Adjusted the Clockwork Stickgun to require Last Tears and Cogs in its crafting recipe.
- Adjusted the Clockwork Stickgun to fire all sticks, not just Sticks of Tears.
- Adjusted the crafting recipe for the Net of Tears to no longer require bottled water.
- Changed the AI of Lunar Sticks and Sticks of Tears to obey gravity.
- Fixed a bug where critters dropped Last Tears.
- Fixed a bug where the Moon Lord's Leech dropped Last Tears.
##1.3.3
- Fixed a bug where the Bleached Stickgun was given the wrong crafting recipe.
##1.3.2
- Fixed a bug where the Bleached Stickgun was uncraftable.
##1.3.0
- Renamed the Emotional Stickgun to the Bleached Stickgun.
- Changed the texture of the Bleached Stickgun.
- Added the Net of Tears.
- Increased the chance for Last Tears to drop.
- Adjusted the recipe for Sticks of Tears.
- Last Tears can now be used to grant the Mourning debuff.
##1.2.0
- Added the Mourning debuff.
- Added Last Tears.
- Tweaked the Emotional Stickgun drop chance from boss bags.
##1.1.0
- Renamed the Emotional Stick to the Stick of Tears.
- Changed the description of the Emotional Stickgun.
##1.0.0
- Restructured the code.
- Added Spooky Sticks.
- Added the Emotional Stickgun and Emotional Sticks.
##0.3.0
- Added a description as well as an icon for the mod browser.
##0.2.0
- Added Stickguns, Sticks, Boreal Sticks, Pearlsticks, Ebonsticks, Shadesticks, Luminite Sticks, and Greedy Sticks.
##0.1.0
- Initial release.
- Added the Big Stick.

Items/MiniStick.cs
using System;
using Terraria;
using Terraria.ID;
using Terraria.ModLoader;
using Terraria.DataStructures;
using System.Collections.Generic;
using Microsoft.Xna.Framework;

namespace ABigStick.Items {
    public class MiniStick : ModProjectile {
        public override void SetStaticDefaults() {
            DisplayName.SetDefault("Stick");
        }

        public override void SetDefaults() {
            projectile.width = 20;
            projectile.height = 20;
            projectile.aiStyle = 1;
            projectile.friendly = true;
			projectile.tileCollide = true;
			projectile.ignoreWater = false;
            projectile.ranged = true;
        }
    }

    public class MiniStickCr2 : ModProjectile {
        public override void SetStaticDefaults() {
            DisplayName.SetDefault("Shadestick");
        }

        public override void SetDefaults() {
            projectile.width = 20;
            projectile.height = 20;
            projectile.aiStyle = 1;
            projectile.friendly = true;
			projectile.tileCollide = true;
			projectile.ignoreWater = false;
            projectile.ranged = true;
        }

		public override void OnHitNPC(NPC target, int damage, float knockback, bool cri) {
		    target.AddBuff(BuffID.Ichor, 5 * 60);
		}
    }

    public class MiniStickCo2 : ModProjectile {
        public override void SetStaticDefaults() {
            DisplayName.SetDefault("Ebonstick");
        }

        public override void SetDefaults() {
            projectile.width = 20;
            projectile.height = 20;
            projectile.aiStyle = 1;
            projectile.friendly = true;
			projectile.tileCollide = true;
			projectile.ignoreWater = false;
            projectile.ranged = true;
        }

		public override void OnHitNPC(NPC target, int damage, float knockback, bool cri) {
		    target.AddBuff(BuffID.CursedInferno, 5 * 60);
		}
    }

    public class MiniStickH2 : ModProjectile {
        public override void SetStaticDefaults() {
            DisplayName.SetDefault("Pearlstick");
        }

        public override void SetDefaults() {
            projectile.width = 20;
            projectile.height = 20;
            projectile.aiStyle = 1;
            projectile.friendly = true;
			projectile.tileCollide = true;
			projectile.ignoreWater = false;
            projectile.ranged = true;
        }

		public override void OnHitNPC(NPC target, int damage, float knockback, bool cri) {
		    target.AddBuff(BuffID.Slow, 5 * 60);
		}
    }

    public class MiniStickBo2 : ModProjectile {
        public override void SetStaticDefaults() {
            DisplayName.SetDefault("Boreal Stick");
        }

        public override void SetDefaults() {
            projectile.width = 20;
            projectile.height = 20;
            projectile.aiStyle = 1;
            projectile.friendly = true;
			projectile.tileCollide = true;
			projectile.ignoreWater = false;
            projectile.ranged = true;
        }

		public override void OnHitNPC(NPC target, int damage, float knockback, bool cri) {
		    target.AddBuff(BuffID.Chilled, 5 * 60);
		}
    }

    public class MiniStickL2 : ModProjectile {
        public override void SetStaticDefaults() {
            DisplayName.SetDefault("Luminite Stick");
        }

        public override void SetDefaults() {
            projectile.width = 20;
            projectile.height = 20;
            projectile.aiStyle = 1;
            projectile.friendly = true;
			projectile.tileCollide = true;
			projectile.ignoreWater = false;
            projectile.ranged = true;
            projectile.penetrate = 999;
        }

		public override void OnHitNPC(NPC target, int damage, float knockback, bool cri) {
		    projectile.damage = projectile.damage + 5;
		}
    }
    
    public class MiniStickD2 : ModProjectile {
        public override void SetStaticDefaults() {
            DisplayName.SetDefault("Rainbow Stick");
        }

        public override void SetDefaults() {
            projectile.width = 20;
            projectile.height = 20;
            projectile.aiStyle = 1;
            projectile.friendly = true;
			projectile.tileCollide = false;
			projectile.ignoreWater = true;
            projectile.ranged = true;
            projectile.penetrate = 999;
        }
    }

    public class MiniStickM2 : ModProjectile {
        public override void SetStaticDefaults() {
            DisplayName.SetDefault("Greedy Stick");
        }

        public override void SetDefaults() {
            projectile.width = 20;
            projectile.height = 20;
            projectile.aiStyle = 1;
            projectile.friendly = true;
			projectile.tileCollide = true;
			projectile.ignoreWater = false;
            projectile.ranged = true;
            projectile.penetrate = 3;
        }

    	public override void OnHitNPC(NPC target, int damage, float knockback, bool cri) {
            int value = Main.rand.Next(20, 200);
            
            if ((Main.rand.NextFloat() >= .45f) && target.type != NPCID.TargetDummy) {
                if (value >= 100) {
                    double partialValue = (double)value/(double)100;
                    int roundedValue = (int)Math.Floor(partialValue);
                    Item.NewItem(target.getRect(), ItemID.SilverCoin, roundedValue);
                    Item.NewItem(target.getRect(), ItemID.CopperCoin, (int)(partialValue-roundedValue));
                } else {
                    Item.NewItem(target.getRect(), ItemID.CopperCoin, value);
                }
            }
		}
    }

    public class MiniStick2 : ModItem {
		public override void SetStaticDefaults() {
			DisplayName.SetDefault("Stick");
		}

        public override void SetDefaults() {
			item.damage = 3;
			item.ranged = true;
			item.width = 20;
			item.height = 20;
			item.consumable = true;
			item.knockBack = 1f;
			item.value = 10;
            item.shoot = mod.ProjectileType("MiniStick");
            item.ammo = item.type;
            item.maxStack = 999;
        }

        public override void AddRecipes() {
            ModRecipe recipe = new ModRecipe(mod);
			recipe.AddIngredient(ItemID.Wood, 50);
            recipe.SetResult(this, 150);
            recipe.AddRecipe();
            ModRecipe recipe3 = new ModRecipe(mod);
			recipe3.AddIngredient(ItemID.PalmWood, 50);
            recipe3.SetResult(this, 150);
            recipe3.AddRecipe();
            ModRecipe recipe4 = new ModRecipe(mod);
			recipe4.AddIngredient(ItemID.RichMahogany, 50);
            recipe4.SetResult(this, 150);
            recipe4.AddRecipe();
        }
    }

    public class MiniStickH : ModItem {
		public override void SetStaticDefaults() {
			DisplayName.SetDefault("Pearlstick");
            Tooltip.SetDefault("Inflicts Slow");
		}

        public override void SetDefaults() {
			item.damage = 5;
			item.ranged = true;
			item.width = 20;
			item.height = 20;
			item.consumable = true;
			item.knockBack = 1f;
			item.value = 10;
            item.shoot = mod.ProjectileType("MiniStickH2");
            item.ammo = mod.ItemType("MiniStick2");
            item.maxStack = 999;
            item.rare = 4;
        }

        public override void AddRecipes() {
            ModRecipe recipe = new ModRecipe(mod);
			recipe.AddIngredient(ItemID.Pearlwood, 50);
            recipe.SetResult(this, 150);
            recipe.AddRecipe();
        }
    }

    public class MiniStickCo : ModItem {
		public override void SetStaticDefaults() {
			DisplayName.SetDefault("Ebonstick");
            Tooltip.SetDefault("Inflicts Cursed Inferno");
		}

        public override void SetDefaults() {
			item.damage = 5;
			item.ranged = true;
			item.width = 20;
			item.height = 20;
			item.consumable = true;
			item.knockBack = 1f;
			item.value = 100;
            item.shoot = mod.ProjectileType("MiniStickCo2");
            item.ammo = mod.ItemType("MiniStick2");
            item.maxStack = 999;
            item.rare = 2;
        }

        public override void AddRecipes() {
            ModRecipe recipe = new ModRecipe(mod);
			recipe.AddIngredient(ItemID.Ebonwood, 50);
            recipe.SetResult(this, 150);
            recipe.AddRecipe();
        }
    }

    public class MiniStickCr : ModItem {
		public override void SetStaticDefaults() {
			DisplayName.SetDefault("Shadestick");
            Tooltip.SetDefault("Inflicts Ichor");
		}

        public override void SetDefaults() {
			item.damage = 7;
			item.ranged = true;
			item.width = 20;
			item.height = 20;
			item.consumable = true;
			item.knockBack = 1f;
			item.value = 100;
            item.shoot = mod.ProjectileType("MiniStickCr2");
            item.ammo = mod.ItemType("MiniStick2");
            item.maxStack = 999;
            item.rare = 2;
        }

        public override void AddRecipes() {
            ModRecipe recipe = new ModRecipe(mod);
			recipe.AddIngredient(ItemID.Shadewood, 50);
            recipe.SetResult(this, 150);
            recipe.AddRecipe();
        }
    }

    public class MiniStickBo : ModItem {
		public override void SetStaticDefaults() {
			DisplayName.SetDefault("Boreal Stick");
            Tooltip.SetDefault("Inflicts Chilled");
		}

        public override void SetDefaults() {
			item.damage = 3;
			item.ranged = true;
			item.width = 20;
			item.height = 20;
			item.consumable = true;
			item.knockBack = 1f;
			item.value = 100;
            item.shoot = mod.ProjectileType("MiniStickBo2");
            item.ammo = mod.ItemType("MiniStick2");
            item.maxStack = 999;
        }

        public override void AddRecipes() {
            ModRecipe recipe = new ModRecipe(mod);
			recipe.AddIngredient(ItemID.BorealWood, 50);
            recipe.SetResult(this, 150);
            recipe.AddRecipe();
        }
    }

    public class MiniStickL : ModItem {
		public override void SetStaticDefaults() {
			DisplayName.SetDefault("Luminite Stick");
            Tooltip.SetDefault("Increases in damage the more targets it pierces through\n'This is irrational.'");
		}

        public override void SetDefaults() {
			item.damage = 15;
			item.ranged = true;
			item.width = 20;
			item.height = 20;
			item.consumable = true;
			item.knockBack = 4f;
			item.value = 0;
            item.rare = 9;
            item.shoot = mod.ProjectileType("MiniStickL2");
            item.ammo = mod.ItemType("MiniStick2");
            item.crit = 30;            
            item.maxStack = 999;
        }

        public override void AddRecipes() {
            ModRecipe recipe = new ModRecipe(mod);
			recipe.AddIngredient(ItemID.LunarBar, 1);
            recipe.SetResult(this, 333);
            recipe.AddTile(TileID.LunarCraftingStation);
            recipe.AddRecipe();
        }
    }
    

    public class MiniStickD : ModItem {
		public override void SetStaticDefaults() {
			DisplayName.SetDefault("Rainbow Stick");
            Tooltip.SetDefault("'Physics? What's that?'");
            Main.RegisterItemAnimation(item.type, new DrawAnimationVertical(5, 7));
		}

        public override void SetDefaults() {
			item.damage = 9999;
			item.ranged = true;
			item.width = 20;
			item.height = 20;
			item.consumable = true;
			item.knockBack = 4f;
			item.value = 9999999;
            item.rare = 11;
            item.expert = true;
            item.shoot = mod.ProjectileType("MiniStickD2");
            item.ammo = mod.ItemType("MiniStick2");
            item.crit = 30;
            item.maxStack = 999;
        }
    }

    public class MiniStickM : ModItem {
		public override void SetStaticDefaults() {
			DisplayName.SetDefault("Greedy Stick");
            Tooltip.SetDefault("45% chance to steal money on impact");
		}

        public override void SetDefaults() {
			item.damage = 7;
			item.ranged = true;
			item.width = 20;
			item.height = 20;
			item.consumable = true;
			item.knockBack = 0f;
			item.value = 200;
            item.shoot = mod.ProjectileType("MiniStickM2");
            item.ammo = mod.ItemType("MiniStick2");
            item.crit = 50;
            item.maxStack = 999;
        }
    }

	public class PirateStickDrop : GlobalNPC {
		public override void NPCLoot(NPC npc) {
            int count = Main.rand.Next(5, 10);
            if(Main.rand.NextBool(Main.expertMode ? 2 : 1, 5)) {
                count = count * 2;
            }
            if (NPC.downedTowerVortex) {
                if (((npc.type >= 212 && npc.type <= 216) || npc.type == 229 || npc.type == 252)) { // all pirates and parrots
                    if (Main.rand.NextFloat() >= .25f) {
                        Item.NewItem(npc.getRect(), mod.ItemType("MiniStickM"), count);
                    }
                } else if (npc.type == 491) { // pirate ship
                    Item.NewItem(npc.getRect(), mod.ItemType("MiniStickM"), count*2);
                }
            }
		}
    }   
}